### PR TITLE
Memory and smart pointers

### PR DESCRIPTION
**Pull request by andresailer**
BEGINRELEASENOTES
- Use smart pointers to wrap objects and simplify memory management.

ENDRELEASENOTES

### CODE REVIEW

File: include/Cell.h [hunk 1]
```diff
@@ -22,64 +22,68 @@
...
-  double getGradient() { return m_gradient; }
+  double getGradient() const { return m_gradient; }
   void setGradient(double gradient) { m_gradient = gradient; }
-  double                  getGradientRZ() { return m_gradientRZ; }
+  double                  getGradientRZ() const { return m_gradientRZ; }
 
   // Angle between two cells. This is assumed to be less than 90 degrees
-  double getAngle(Cell* cell2) {
+  double getAngle(SCell cell2) const { return getAngle(cell2.get()); }
+  double getAngle(Cell* cell2) const {
```
Comment (dhynds) [inline include/Cell.h]: to be removed and all functionality moved into shared pointer function?

File: include/Cell.h [hunk 2]
```diff
@@ -22,64 +22,68 @@
...
     return fabs(std::atan((cell2->getGradient() - m_gradient) / (1 + m_gradient * cell2->getGradient())));
   }
-  double getAngleRZ(Cell* cell2) {
+  double getAngleRZ(SCell cell2) const { return getAngleRZ(cell2.get()); }
+  double getAngleRZ(Cell* cell2) const {
```
Comment (dhynds) [inline include/Cell.h]: same comment as above

File: include/Cell.h [hunk 3]
```diff
@@ -22,64 +22,68 @@
...
+  void update(SCell const& cell2) {
     if ((cell2->getWeight() + 1) > m_weight)
       m_weight = cell2->getWeight() + 1;
   }
 
   // The cell has a memory of all cells that connect to it, and all cells that it connects to. If several cells point to this
   // cell, then the weight taken from the highest weighted of those (longest chain)
-  void setFrom(Cell* cell2) {
-    m_from.push_back(cell2);
+  void setFrom(SCell const& cell2) {
```
Comment (dhynds) [inline include/Cell.h]: what is the purpose of the const here?

File: include/Cell.h [hunk 4]
```diff
@@ -22,64 +22,68 @@
...
+    m_from.push_back(WCell(cell2));
     m_weights.push_back(cell2->getWeight());
     if ((cell2->getWeight() + 1) > m_weight)
       m_weight = cell2->getWeight() + 1;
   }
-  void setTo(Cell*    cell2) { m_to.push_back(cell2); }
-  std::vector<Cell*>* getFrom() { return &m_from; }
-  std::vector<Cell*>* getTo() { return &m_to; }
+  void setTo(SCell const& cell2) { m_to.push_back(WCell(cell2)); }
+  WeakCells*              getFrom() { return &m_from; }
```
Comment (dhynds) [inline include/Cell.h]: raw pointer to vector of shared pointers? Should this also be a shared pointer? Or non-pointer?

File: include/Cell.h [hunk 5]
```diff
@@ -22,64 +22,68 @@
...
+  WeakCells*              getTo() { return &m_to; }
 
-  double doca() {
+  double doca() const {
```
Comment (dhynds) [inline include/Cell.h]: what function does const perform here?

File: include/ConformalTracking.h [hunk 6]
```diff
@@ -64,44 +64,44 @@ class ConformalTracking : public Processor {
...
   // Plotting function for displaying cells
-  void drawline(KDCluster*, KDCluster*, int, int style = 1);
+  void drawline(SKDCluster const&, SKDCluster const&, int, int style = 1);
 
   // Pattern recognition algorithms:
 
   // Cell creation
-  KDCluster* extrapolateCell(Cell*, double);
-  void       extendSeedCells(std::vector<Cell*>&, KDTree*, bool, const std::vector<KDCluster*>&);
+  SKDCluster extrapolateCell(Cell::SCell const&, double);
```
Comment (dhynds) [inline include/ConformalTracking.h]: what is the purpose of the const&? The cell that is pointed to will be modified during this function

File: include/ConformalTracking.h [hunk 7]
```diff
@@ -64,44 +64,44 @@ class ConformalTracking : public Processor {
...
+  void       extendSeedCells(SharedCells&, KDTree*, bool, const SharedKDClusters&);
 
   // Track finding
-  void buildNewTracks(std::vector<KDTrack*>&, std::vector<KDCluster*>&, KDTree*, bool radialSearch = false);
-  void extendTracks(std::vector<KDTrack*>&, std::vector<KDCluster*>&, KDTree*);
-  void combineCollections(std::vector<KDCluster*>&, KDTree*&, std::vector<int>, std::map<int, std::vector<KDCluster*>>);
+  void buildNewTracks(UniqueKDTracks&, SharedKDClusters&, KDTree*, bool radialSearch = false);
```
Comment (dhynds) [inline include/ConformalTracking.h]: Should one also make the KDTree a unique/shared pointer ?

File: include/KDTrack.h [hunk 8]
```diff
@@ -61,9 +61,9 @@ class KDTrack {
   double chi2ndofZS() { return m_chi2ndofZS; }
 
   // Clusters and kalman track pointer
-  int                     nPoints() { return m_nPoints; }
-  std::vector<KDCluster*> clusters() { return m_clusters; }
-  KalmanTrack*            kalmanTrack() { return m_kalmanTrack; }
+  int              nPoints() { return m_nPoints; }
+  SharedKDClusters clusters() { return m_clusters; }
+  KalmanTrack*     kalmanTrack() { return m_kalmanTrack; }
```
Comment (dhynds) [inline include/KDTrack.h]: Should this not be shared/unique pointer?

File: include/Cell.h (hunk 3 above)
Comment (andresailer) [reply include/Cell.h]: This signals that cell2 is not modified in the function call, as we are dealing with a shared_ptr that we copy we would probably have the same performance if we just pass by copy here.

File: include/KalmanTrack.h [hunk 9]
```diff
@@ -74,11 +74,11 @@ class KalmanTrack {
   ~KalmanTrack() = default;
 
   // Functions
-  double addCluster(KDCluster*);  // returns delta chi2 for adding this cluster
+  double addCluster(SKDCluster);  // returns delta chi2 for adding this cluster
 
   // Member variables
-  KDTrack*                m_conformalTrack = nullptr;  // The parent conformal track
-  std::vector<KDCluster*> m_kalmanClusters{};          // The additional clusters to which the kalman filter is applied
+  KDTrack*         m_conformalTrack = nullptr;  // The parent conformal track
```
Comment (dhynds) [inline include/KalmanTrack.h]: Why is this left as a raw pointer?

File: src/ConformalTracking.cc [hunk 10]
```diff
@@ -1085,9 +1074,9 @@ void ConformalTracking::buildNewTracks(std::vector<KDTrack*>& conformalTracks, s
     // We create all acceptable tracks by looping over all cells with high enough weight to create
     // a track and trace their route back to the seed hit. We then have to choose the best candidate
     // at the end (by minimum chi2 of a linear fit)
-    std::map<Cell*, bool> usedCells;
-    std::map<Cell*, bool> usedCells2;
-    std::vector<KDTrack*> cellTracks;
+    std::map<Cell::SCell, bool> usedCells;
```
Comment (dhynds) [inline src/ConformalTracking.cc]: Why Cell::SCell? Is there another SCell?

File: src/ConformalTracking.cc [hunk 11]
```diff
@@ -1106,23 +1095,20 @@ void ConformalTracking::buildNewTracks(std::vector<KDTrack*>& conformalTracks, s
...
 
-      for (size_t itTrack = 0; itTrack < candidateTracksTemp.size(); itTrack++) {
-        if (int(candidateTracksTemp[itTrack]->size()) >= (m_minClustersOnTrack - 1))
-          candidateTracks.push_back(candidateTracksTemp[itTrack]);
-        else
-          delete candidateTracksTemp[itTrack];
-      }
+      copy_if(std::make_move_iterator(candidateTracksTemp.begin()), std::make_move_iterator(candidateTracksTemp.end()),
+              std::back_inserter(candidateTracks),
+              [this](UcellularTrack const& track) { return (int(track->size()) >= (m_minClustersOnTrack - 1)); });
```
Comment (dhynds) [inline src/ConformalTracking.cc]: WTF?!

File: src/ConformalTracking.cc [hunk 12]
```diff
@@ -1106,23 +1095,20 @@ void ConformalTracking::buildNewTracks(std::vector<KDTrack*>& conformalTracks, s
...
 
       // Debug plotting
       if (m_debugPlots && m_eventNumber == 2) {
         m_canvConformalEventDisplayAcceptedCells->cd();
         for (size_t iTr = 0; iTr < candidateTracks.size(); iTr++) {
-          cellularTrack* track = candidateTracks[iTr];
+          cellularTrack* track = candidateTracks[iTr].get();
```
Comment (dhynds) [inline src/ConformalTracking.cc]: Why use raw pointers here?

File: src/ConformalTracking.cc [hunk 13]
```diff
@@ -1613,22 +1583,21 @@ void ConformalTracking::extendHighPT(std::vector<KDTrack*>& conformalTracks, std
...
     // Now have all possible extensions. Get the lowest chi2
-    double   lowestChi2ndof  = 10000000.;
-    KDTrack* lowestChi2Track = NULL;
+    double   lowestChi2ndof = 10000000.;
+    UKDTrack lowestChi2Track;
     for (size_t itTrack = 0; itTrack < extensions.size(); itTrack++) {
       // Get the extension, add it to the original hit
-      cellularTrack*     extension = extensions[itTrack];
-      vector<KDCluster*> hits      = track->clusters();
+      cellularTrack*   extension = extensions[itTrack].get();
```
Comment (dhynds) [inline src/ConformalTracking.cc]: again, why raw pointers?

File: include/ConformalTracking.h (hunk 6 above)
Comment (dhynds) [reply include/ConformalTracking.h]: nevermind, thats not true...

File: src/ConformalTracking.cc (hunk 12 above)
Comment (dhynds) [reply src/ConformalTracking.cc]: use auto ;-)

File: include/Cell.h (hunk 4 above)
Comment (andresailer) [reply include/Cell.h]: Reference

File: include/Cell.h (hunk 5 above)
Comment (andresailer) [reply include/Cell.h]: Help the compiler optimize the code to signal that the function does not modify any of the member variables

File: include/ConformalTracking.h (hunk 7 above)
Comment (andresailer) [reply include/ConformalTracking.h]: Done

File: src/ConformalTracking.cc (hunk 11 above)
Comment (andresailer) [reply src/ConformalTracking.cc]: That is a strange way to spell `beautiful` :)

File: src/ConformalTracking.cc (hunk 12 above)
Comment (andresailer) [reply src/ConformalTracking.cc]: Done

**Pull request merged**